File: rs/anda_cognitive_nexus/src/attention/catalog.rs
use super::*;
use anda_db::query::{Filter, RangeQuery};
use anda_db_schema::Fv;

/// One bounded discovery scan. A complete page exhausts its authorized snapshot,
/// not processing obligations. Claims must re-read current versions and fences.
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct WakePage {
    pub items: Vec<WakeRecord>,
    pub snapshot_seq: u64,
    pub scanned: usize,
    pub next_cursor: Option<String>,
    pub complete: bool,
}

#[derive(Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
struct Cursor {
    format: String,
    space: String,
    principal: String,
    instance: Option<String>,
    basis: String,
    snapshot_seq: u64,
    after_id: u64,
}

impl Session {
    /// Read-only, index-backed pagination over native wake history. Empty
    /// intermediate pages are valid; callers must follow next_cursor.
    pub async fn list_wakes(
        &self,
        space: &str,
        cursor: Option<&str>,
        scan_limit: usize,
    ) -> Result<WakePage, KipError> {
        if !(1..=200).contains(&scan_limit) {
            return Err(invalid("wake scan limit must be 1..=200"));
        }
        let _guard = self.nexus.read_guard().await?;
        let authority = self.effective_authority(space).await?;
        authority
            .authorize(
                Permission::Maintain,
                &ResourceContext::default(),
                &self.auth,
            )
            .into_result()?;
        let cx =
            Context::open(&self.nexus.store, space, None, None, &authority, &self.auth).await?;
        let current_basis = digest(&basis(&cx))?;
        let instance = self
            .nexus
            .store
            .control_at(space, CONFIG, u64::MAX)
            .await?
            .and_then(|r| {
                r.value["scope"]["space_instance"]
                    .as_str()
                    .map(str::to_owned)
            });
        let mut position = if let Some(cursor) = cursor {
            if cursor.len() > 8192 {
                return Err(invalid("wake cursor exceeds bound"));
            }
            let bytes = hex::decode(cursor).map_err(|_| invalid("invalid wake cursor"))?;
            serde_json::from_slice::<Cursor>(&bytes).map_err(|_| invalid("invalid wake cursor"))?
        } else {
            Cursor {
                format: "nexus:wake-cursor-v1".into(),
                space: space.into(),
                principal: self.auth.principal_id.clone(),
                instance: instance.clone(),
                basis: current_basis.clone(),
                snapshot_seq: cx.pinned_seq,
                after_id: 0,
            }
        };
        if position.format != "nexus:wake-cursor-v1"
            || position.space != space
            || position.principal != self.auth.principal_id
            || position.instance != instance
            || position.basis != current_basis
            || position.snapshot_seq > cx.pinned_seq
            || position.after_id > anda_kip::MAX_SAFE_INTEGER
        {
            return Err(conflict("wake_cursor_basis_changed"));
        }
        let table = self.nexus.store.control_records();
        let ids = table
            .query_ids(
                Filter::And(vec![
                    Box::new(crate::store::eq_field("space", Fv::Text(space.into()))),
                    Box::new(crate::store::eq_field("kind", Fv::Text("wake".into()))),
                    Box::new(Filter::Field((
                        "seq".into(),
                        RangeQuery::Le(Fv::U64(position.snapshot_seq)),
                    ))),
                    Box::new(Filter::Field((
                        "_id".into(),
                        RangeQuery::Gt(Fv::U64(position.after_id)),
                    ))),
                ]),
                Some(scan_limit + 1),
            )
            .await
            .map_err(crate::error::db_error)?;
        let complete = ids.len() <= scan_limit;
        let scanned = ids.len().min(scan_limit);
        let mut items = Vec::new();
        for id in ids.into_iter().take(scan_limit) {
            position.after_id = id;
            let row: ControlRecordRow = table.get_as(id).await.map_err(crate::error::db_error)?;
            // Native controls append versions in commit order. Select one head
            // with a bounded index lookup, not an unbounded history read per row.
            let latest = table
                .query_last_ids(
                    Filter::And(vec![
                        Box::new(crate::store::eq_field("space", Fv::Text(space.into()))),
                        Box::new(crate::store::eq_field("key", Fv::Text(row.key.clone()))),
                        Box::new(Filter::Field((
                            "seq".into(),
                            RangeQuery::Le(Fv::U64(position.snapshot_seq)),
                        ))),
                    ]),
                    Some(1),
                )
                .await
                .map_err(crate::error::db_error)?;
            if latest.first() != Some(&id) {
                continue;
            }
            let wake: WakeRecord =
                serde_json::from_value(row.value).map_err(|_| invalid("corrupt wake record"))?;
            if wake.wake_ref != row.key || wake.version != row.version {
                return Err(invalid("wake identity mismatch"));
            }
            let mut visible = true;
            for reference in [&wake.fire.watch_ref, &wake.fire_activity_ref] {
                let source = self.nexus.store.get_element(reference.parse()?).await?;
                visible &= source.space() == space
                    && authority
                        .may_read(&source, &self.auth)
                        .is_some_and(|v| v.content && v.constraints.fields.is_empty());
            }
            if visible {
                items.push(wake);
            }
        }
        Ok(WakePage {
            items,
            snapshot_seq: position.snapshot_seq,
            scanned,
            next_cursor: if complete {
                None
            } else {
                Some(hex::encode(
                    serde_json::to_vec(&position).map_err(|e| invalid(&e.to_string()))?,
                ))
            },
            complete,
        })
    }
}

File: ts/kip-do/src/attention/catalog.ts
import type { Session } from '../nexus.js'
import type { ControlRecord } from '../control.js'
import type { WakePage, WakeRecord, AttentionConfig } from './types.js'
import { authorizedArtifact, publishControl } from '../control.js'
import { errors } from '../errors.js'
import { digest } from '../schema/contracts.js'
import { parseElementId } from '../id.js'
import {
  CONFIG,
  obj,
  json,
  eq,
  bytes,
  safe,
  invalid,
  conflict,
  permit,
  fullRead,
  context,
  basis,
  validateConfig,
} from './common.js'
import { loadWake } from './work.js'

export function setAttentionConfig(
  s: Session,
  space: string,
  expected: number,
  config: AttentionConfig,
): import('../json.js').JsonMap {
  validateConfig(config)
  const previous = s.nexus.store.controlAt(space, CONFIG)
  if (previous && !eq(obj(previous.value).scope, config.scope))
    invalid('attention instance cannot be replaced in place')
  const row = publishControl(
    s.nexus.store,
    space,
    CONFIG,
    'policy',
    expected,
    json(config),
    { principal_id: s.auth.principal_id },
  )
  return { version: row.version, config: row.value }
}
export function authorizeControlRead(
  s: Session,
  space: string,
  row: ControlRecord,
): void {
  if (row.kind === 'wake') loadWake(s, space, row.key)
  else if (row.kind === 'dispatch' && row.key.startsWith('dispatch/v1/')) {
    const request = obj(obj(row.value).request)
    loadWake(s, space, String(request.wake_ref))
    fullRead(s, space, String(request.attempt_ref))
  } else if (row.kind === 'runtime' && row.key !== CONFIG) {
    if (!row.key.startsWith('watch-evaluation/v1/'))
      throw errors.notAuthorized(
        'attention runtime records use their dedicated read or replay API',
      )
    authorizedArtifact(
      s.nexus.store,
      space,
      String(obj(obj(row.value).material).artifact_ref),
      s.effectiveAuthority(space),
      s.auth,
    )
  }
}
interface Cursor {
  format: string
  space: string
  principal: string
  instance: string | null
  basis: string
  snapshot_seq: number
  after_id: number
}
export function listWakes(
  s: Session,
  space: string,
  cursor: string | null,
  limit: number,
): WakePage {
  if (!Number.isInteger(limit) || limit < 1 || limit > 200)
    invalid('wake scan limit must be 1..=200')
  permit(s, space, 'maintain')
  const store = s.nexus.store,
    authority = s.effectiveAuthority(space),
    head = store.currentSeq(space),
    currentBasis = digest(basis(context(s, space)))
  const instance = (obj(obj(store.controlAt(space, CONFIG)?.value).scope)
    .space_instance ?? null) as string | null
  let position: Cursor = {
    format: 'nexus:wake-cursor-v1',
    space,
    principal: s.auth.principal_id,
    instance,
    basis: currentBasis,
    snapshot_seq: head,
    after_id: 0,
  }
  if (cursor !== null) {
    if (bytes(cursor) > 8192 || !/^(?:[0-9a-f]{2})+$/.test(cursor))
      invalid('invalid wake cursor')
    try {
      position = JSON.parse(
        new TextDecoder('utf-8', { fatal: true, ignoreBOM: false }).decode(
          Uint8Array.from(cursor.match(/../g)!, (c) => parseInt(c, 16)),
        ),
      ) as Cursor
    } catch {
      invalid('invalid wake cursor')
    }
  }
  if (
    !position ||
    position.format !== 'nexus:wake-cursor-v1' ||
    position.space !== space ||
    position.principal !== s.auth.principal_id ||
    position.instance !== instance ||
    position.basis !== currentBasis ||
    !safe(position.snapshot_seq) ||
    position.snapshot_seq > head ||
    !safe(position.after_id)
  )
    conflict('wake_cursor_basis_changed')
  const rows = store.all<ControlRecord>(
    'kip_control_records',
    "SELECT * FROM kip_control_records WHERE space = ? AND kind = 'wake' AND seq <= ? AND id > ? ORDER BY id LIMIT ?",
    space,
    position.snapshot_seq,
    position.after_id,
    limit + 1,
  )
  const complete = rows.length <= limit,
    scanned = Math.min(rows.length, limit),
    items: WakeRecord[] = []
  for (const row of rows.slice(0, limit)) {
    position.after_id = row.id
    if (store.controlAt(space, row.key, position.snapshot_seq)?.id !== row.id)
      continue
    const wake = row.value as unknown as WakeRecord
    if (wake.wake_ref !== row.key || wake.version !== row.version)
      invalid('wake identity mismatch')
    if (
      [wake.fire.watch_ref, wake.fire_activity_ref].every((ref) => {
        const e = store.load(parseElementId(ref)),
          v = e && authority.mayRead(e, s.auth)
        return (
          e?.row.space === space && v?.content && !v.constraints.fields.length
        )
      })
    )
      items.push(wake)
  }
  const nextCursor = complete
    ? null
    : [...new TextEncoder().encode(JSON.stringify(position))]
        .map((b) => b.toString(16).padStart(2, '0'))
        .join('')
  return {
    items,
    snapshot_seq: position.snapshot_seq,
    scanned,
    next_cursor: nextCursor,
    complete,
  }
}
